frontend/src/app/modules/users/users-selectors.js

```diff
@@ -1,14 +1,68 @@
 import { createSelector } from 'reselect'
 
+import * as Filters from '../../const/leaderboard-filters'
+import { didUserPlayMatch } from '../matches/matches-selectors'
+import {
+  computeLongestWinStreak,
+  computeWinRatio,
+} from '../matches/matches-computations'
+
 export const getUsers = state => state.users
+const getMatches = state => state.matches
+const getFilters = state => state.filters
 
 export const getUser = createSelector(
   getUsers,
   (stats, userId) => userId,
   (users, userId) => users.find(user => user.id === userId),
 )
 
+export const getTopRatedUsers = createSelector(
+  getUsers,
+  users => users.sort((user1, user2) => user2.rating - user1.rating),
+)
+
 export const getTopUsers = createSelector(
   getUsers,
-  users => [...users].sort((user1, user2) => user2.rating - user1.rating),
+  getMatches,
+  getFilters,
+  (users, matches, filters) => sortTopUsers(users, matches, filters),
 )
+
+const sortTopUsers = (users, matches, filters) => {
+  const order = filters.order === Filters.orderTypes.ASC ? -1 : 1
+  const minDate = new Date()
+  minDate.setDate(minDate.getDate() - filters.timespan)
+  const matchesLast = matches
+    .filter(match => filters.timespan === Filters.timespanTypes.AllTime || match.date > minDate)
+    .sort((match1, match2) => match2.date - match1.date)
+
+  const usersWithStatistics  = users.map(user => getUserStatistics(user, matchesLast))
+  switch (filters.criteria) {
+    case Filters.criteriaTypes.Wins:
+      return usersWithStatistics.sort((u1, u2) => (u2.stats.wins - u1.stats.wins) * order)
+    case Filters.criteriaTypes.Ratio:
+      return usersWithStatistics.sort((u1, u2) => (u2.stats.winRatio - u1.stats.winRatio) * order)
+    case Filters.criteriaTypes.Streak:
+      return usersWithStatistics.sort((u1, u2) => (u2.stats.streak - u1.stats.streak) * order)
+    case Filters.criteriaTypes.Matches:
+      return usersWithStatistics.sort((u1, u2) => (u2.stats.matches - u1.stats.matches) * order)
+    default:
+      return usersWithStatistics.sort((u1, u2) => (u2.rating - u1.rating) * order)
+  }
```

Comment:
A little suggestion - in order not to duplicate the `sort` call and the `* order` multiplication, we could rewrite it like this:

```
  const usersWithStatistics  = users.map(user => getUserStatistics(user, matchesLast))

  let valueExtractor;
  switch (filters.criteria) {
    case Filters.criteriaTypes.Wins:
      valueExtractor = u2 => u2.stats.wins;
    case Filters.criteriaTypes.Ratio:
      valueExtractor = u2 => u2.stats.winRatio;
    case Filters.criteriaTypes.Streak:
      valueExtractor = u2 => u2.stats.streak;
    case Filters.criteriaTypes.Matches:
      valueExtractor = u2 => u2.stats.matches;
    default:
      valueExtractor = u2 => u2.stats.rating;
  } 

  return usersWithStatistics.sort((u1, u2) => (valueExtractor(u2) - valueExtractor(u1) * order);
```

Reply:
After [moving points parsing to selector](https://github.com/salsita/foosball-rating/pull/92/commits/66516b36eac73b2ce0f63c758689cac523dc09f9#diff-6bb144976a04c9e3de620e0c0828e32e), I realised that sorting can be easily done only through the parsed points.